semanticdb-javac/src/main/java/com/sourcegraph/semanticdb_javac/SemanticdbSignatures.java

```diff
@@ -0,0 +1,265 @@
+    private final GlobalSymbolsCache cache;
+    private final LocalSymbolsCache locals;
+
+    private SemanticdbTypeVisitor(GlobalSymbolsCache cache, LocalSymbolsCache locals) {
+      this.cache = cache;
+      this.locals = locals;
+    }
+
+    @Override
+    public Semanticdb.Type visitDeclared(DeclaredType t, Void unused) {
+      boolean isExistential =
+          t.getTypeArguments().stream().anyMatch((type) -> type instanceof WildcardType);
+
+      ArrayList<Semanticdb.Type> typeParams = new ArrayList<>();
+      Scope.Builder declarations = Scope.newBuilder();
+      for (TypeMirror type : t.getTypeArguments()) {
+        Semanticdb.Type visited = super.visit(type);
+        if (visited == null) {
+          pprint("NULL TYPE PARAM " + type + " " + type.getClass());
+          continue;
+        }
+        typeParams.add(visited);
+
+        if (type instanceof WildcardType) {
+          TypeSignature.Builder typeSig = TypeSignature.newBuilder();
+          WildcardType wildcardType = (WildcardType) type;
+
+          // semanticdb spec asks for List() not None for  type_parameters field
+          typeSig.setTypeParameters(Scope.newBuilder());
+
+          if (wildcardType.getExtendsBound() != null) {
+            typeSig.setUpperBound(super.visit(wildcardType.getExtendsBound()));
+          } else if (wildcardType.getSuperBound() != null) {
+            typeSig.setLowerBound(super.visit(wildcardType.getSuperBound()));
+          }
+
+          declarations.addHardlinks(
+              SymbolInformation.newBuilder()
+                  .setSymbol("local_wildcard")
+                  .setSignature(Signature.newBuilder().setTypeSignature(typeSig)));
+        } else {
+          declarations.addSymlinks(cache.semanticdbSymbol(((Type) type).asElement(), locals));
+        }
+      }
+
+      if (!isExistential) {
+        return Semanticdb.Type.newBuilder()
+            .setTypeRef(
+                TypeRef.newBuilder()
+                    .setSymbol(cache.semanticdbSymbol(t.asElement(), locals))
+                    .addAllTypeArguments(typeParams))
+            .build();
+      } else {
+        return Semanticdb.Type.newBuilder()
+            .setExistentialType(
+                ExistentialType.newBuilder()
+                    .setTpe(
+                        Semanticdb.Type.newBuilder()
+                            .setTypeRef(
+                                TypeRef.newBuilder()
+                                    .setSymbol(cache.semanticdbSymbol(t.asElement(), locals))
+                                    .addAllTypeArguments(typeParams)))
+                    .setDeclarations(declarations))
+            .build();
+      }
+    }
+
+    @Override
+    public Semanticdb.Type visitArray(ArrayType t, Void unused) {
+      Semanticdb.Type arrayComponentType = super.visit(t.getComponentType());
+      if (arrayComponentType == null) {
+        pprint("ARRAY GAVE NULL " + t.getComponentType() + " " + t);
```

Comment:
Remove?

Reply:
Im wondering if, for cases like this, we should add some magic symbol as with `"local_wildcard"`, but instead `"unresolved"`. I would personally prefer this in an editor e.g. `class Type extends [unresolved]` as opposed to leaving me to believe that it extends nothing and is a correct hover of `class Type`. What do you think?

Reply:
I think it's OK to return a magic symbol instead of silently failing. It would make it easier to diagnose a bug/problem if a user encounters it. I think it's also OK to return null and skip the signature in these cases.

Reply:
https://github.com/sourcegraph/lsif-java/pull/126/commits/dcb3cbbd06245934426cfcd4cff16ddea9b45123 Let me know what you think of this change